homeassistant/components/abode.py

```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+                          if device.entity_id in entity_ids]
+
+        for device in target_devices:
+            device.trigger()
+
+    descriptions = load_yaml_config_file(
+        path.join(path.dirname(__file__), 'services.yaml'))[DOMAIN]
+
+    hass.services.register(
+        DOMAIN, SERVICE_SETTINGS, change_setting,
+        descriptions.get(SERVICE_SETTINGS),
+        schema=CHANGE_SETTING_SCHEMA)
+
+    hass.services.register(
+        DOMAIN, SERVICE_CAPTURE_IMAGE, capture_image,
+        descriptions.get(SERVICE_CAPTURE_IMAGE),
+        schema=CAPTURE_IMAGE_SCHEMA)
+
+    hass.services.register(
+        DOMAIN, SERVICE_TRIGGER, trigger_quick_action,
+        descriptions.get(SERVICE_TRIGGER),
+        schema=TRIGGER_SCHEMA)
+
+
+def setup_hass_events(hass):
+    """Home assistant start and stop callbacks."""
+    def startup(event):
+        """Listen for push events."""
+        if not hass.data[DOMAIN].polling:
+            hass.data[DOMAIN].abode.events.start()
+
     def logout(event):
         """Logout of Abode."""
-        abode.stop_listener()
-        abode.logout()
+        if not hass.data[DOMAIN].polling:
+            hass.data[DOMAIN].abode.events.stop()
+
+        hass.data[DOMAIN].abode.logout()
         _LOGGER.info("Logged out of Abode")
 
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, startup)
     hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout)
 
-    def startup(event):
-        """Listen for push events."""
-        abode.start_listener()
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, startup)
+def setup_abode_events(hass):
+    """Event callbacks."""
+    import abodepy.helpers.timeline as TIMELINE
 
-    return True
+    def event_callback(event_json, event=None):
+        """Handle an event callback from Abode."""
+        if event:
+            data = {
+                ATTR_DEVICE_ID: event_json.get(ATTR_DEVICE_ID, ''),
+                ATTR_DEVICE_NAME: event_json.get(ATTR_DEVICE_NAME, ''),
+                ATTR_DEVICE_TYPE: event_json.get(ATTR_DEVICE_TYPE, ''),
+                ATTR_EVENT_CODE: event_json.get(ATTR_EVENT_CODE, ''),
+                ATTR_EVENT_NAME: event_json.get(ATTR_EVENT_NAME, ''),
+                ATTR_EVENT_TYPE: event_json.get(ATTR_EVENT_TYPE, ''),
+                ATTR_EVENT_UTC: event_json.get(ATTR_EVENT_UTC, ''),
+                ATTR_USER_NAME: event_json.get(ATTR_USER_NAME, ''),
+                ATTR_DATE: event_json.get(ATTR_DATE, ''),
+                ATTR_TIME: event_json.get(ATTR_TIME, ''),
+            }
+            _LOGGER.info(data)
+            hass.bus.fire(event, data)
+
+    events = [TIMELINE.ALARM_GROUP, TIMELINE.ALARM_END_GROUP,
+              TIMELINE.PANEL_FAULT_GROUP, TIMELINE.PANEL_RESTORE_GROUP,
+              TIMELINE.AUTOMATION_GROUP]
+
+    for event in events:
+        hass.data[DOMAIN].abode.events.add_event_callback(
+            event,
+            partial(event_callback, event=event))
```

Comment:
Why not move the event parameter to be the first? that way you can use positional parameters over keyword args

Reply:
IIRC I placed it second so that the event_callback from the underlying library I wrote would correctly pass the callback data into the first parameter. I'm not super familiar with python partials/parameter details so this was mostly just working off of Google. If you could explain a bit more here or on Discord I'd appreciate it.

Reply:
So the idea of partial is that you will auto fill in parameters.

```python
# The method we want to call
def my_callback(first_param, second_param):
    print('first', first_param)
    print('second', second_param)

# This:
def my_wrapper(second_param):
    my_callback('something_for_first', second_param)

# Is equal to:
my_wrapper = partial(my_callback, 'something_for_first')
```